src/kerberos/mod.rs

```diff
@@ -745,56 +736,43 @@ impl<'a> Kerberos {
-                                let mut sha1 = Sha1::new();
-                                sha1.update(data_to_sign);
-                                let hash = sha1.finalize().to_vec();
-
-                                let mut smart_card = SmartCard::new_emulated(
-                                    reader_name.clone().into(),
-                                    pin.clone(),
-                                    private_key_pem.as_ref(),
-                                    certificate.clone(),
-                                )?;
-                                smart_card.sign(hash)
-                            }),
-                            with_pre_auth: false,
-                            authenticator_nonce: OsRng.gen::<[u8; 4]>(),
-                        }))
-                    }
-                };
+                let pa_data_options =
+                    match credentials {
+                        CredentialsBuffers::AuthIdentity(auth_identity) => {
+                            let domain = utf16_bytes_to_utf8_string(&auth_identity.domain);
+                            let salt = format!("{}{}", domain, username);
+
+                            AsReqPaDataOptions::AuthIdentity(GenerateAsPaDataOptions {
+                                password: &password,
+                                salt: salt.as_bytes().to_vec(),
+                                enc_params: self.encryption_params.clone(),
+                                with_pre_auth: false,
+                            })
+                        }
+                        CredentialsBuffers::SmartCard(smart_card) => {
+                            let private_key_pem =
+                                utf16_bytes_to_utf8_string(smart_card.private_key_pem.as_ref().ok_or_else(|| {
+                                    Error::new(ErrorKind::InternalError, "scard private key is missing")
+                                })?);
+                            self.dh_parameters = Some(generate_client_dh_parameters(&mut OsRng)?);
+
+                            AsReqPaDataOptions::SmartCard(Box::new(pk_init::GenerateAsPaDataOptions {
+                                p2p_cert: picky_asn1_der::from_bytes(&smart_card.certificate)?,
+                                kdc_req_body: &kdc_req_body,
+                                dh_parameters: self.dh_parameters.clone().unwrap(),
+                                sign_data: Box::new(move |data_to_sign| {
+                                    let mut sha1 = Sha1::new();
+                                    sha1.update(data_to_sign);
+                                    let hash = sha1.finalize().to_vec();
+                                    let private_key = PrivateKey::from_pem_str(&private_key_pem)?;
+                                    let rsa_private_key = RsaPrivateKey::try_from(&private_key)?;
+                                    Ok(rsa_private_key.sign(Pkcs1v15Sign::new::<Sha1>(), &hash)?)
+                                }),
```

Comment:
I'm concerned a bit with removing the _smartcard_ layer and just using the simple RSA signature generation. But I think it's not a problem for us. Because of two reasons:

* The `winscard` crate does the same.
* Anyway, we don't support system-provided smart cards in our current sspi implementation. But we will in the next releases and I'll look into this code and refactor it.

cc @awakecoding @CBenoit

Reply:
Thank you for the heads-up! If you’re going to look into it afterwards, I’m fine with that :+1:

Reply:
The aim was to cut the dependency on `winscard` crate and to avoid including unused libraries for system-provided smart cards when it's not needed. So yeah, the code is basically copied over from emulated part of `winscard` crate.